### PR TITLE
Find and replace within a single cell

### PR DESCRIPTION
**Pull request by jess-x**
## References

per #2017, this PR addresses the much requested feature of find/replace within a single cell

## User-facing changes
Screen Shot 2021-04-13 at 8 13 05 PM

Screen Recording 2021-04-13 at 8 14 53 PM

### CODE REVIEW

**Comment by blink1073** [comment]
Thanks for working on this one @jess-x, code changes look good so far!

File: packages/documentsearch/src/interfaces.ts [hunk 1]
```diff
@@ -29,6 +29,11 @@ export interface IDisplayState {
    */
   useRegex: boolean;
 
+  /**
+   * Should the search be within single cell?
+   */
+  searchCell: boolean;
```
Comment (goanpeca) [inline packages/documentsearch/src/interfaces.ts:35]: How about something like
Select if the search happens on the entire document or just a single cell
`Should the search be within single cell?`

Also `searchCell` could be `searchSingleCell` or `searchActiveCell` to make it more descriptiive?

Thoughts @blink1073 ?

**Review by goanpeca — APPROVED** [review]
Looks great @jess-x Thanks!

Left some comment on the parameter name and doc.

**Comment by jasongrout** [comment]
Thanks for working on this!

Testing the current PR, I noticed a few things:

First, there already is a place for changing the scope of the search: the triple-dot menu on the left opens up a dropdown letting you decide about searching outputs as well as inputs. You can see it here:

Screen Shot 2021-04-12 at 8 55 35 AM

We originally decided to make this "scope" area a separate dropdown so that it could easily be extended. Since the search works for any document type (not just notebooks), we needed a way for each document to define its own scoping options. For example, if you are searching a geojson file represented as a map, you may want the scope to be the current visible section of the map, or the current selection on the map, etc. And you may want different scope options to be toggled independent of each other, like in the notebook you may want to toggle searching inputs vs outputs, but separately toggle single cell or whole notebook. For this reason, I'd suggest putting the scope option down in the existing scope selection area, by the the "Search cell outputs" scope option.


Second, it seems that the search is not updated when a new cell is selected. I would imagine that if I searched in a cell, then changed the cell, the search should be cleared from the old cell and redone in the new cell. As it is now, if I select a new cell after toggling the cell scope on, I can get into a situation like this, showing that I am seeing the second occurence out of one occurrence, and both being highlighted in different cells.

Screen Shot 2021-04-12 at 8 56 23 AM

**Comment by jasongrout** [comment]
Reading the code at https://github.com/jupyterlab/jupyterlab/blob/b527a2c9753cd7b2ba1db79c890baa79ce9c2751/packages/documentsearch/src/searchoverlay.tsx#L317 a bit more, it seems that the code isn't very generic about new scope selections, so it seems like the ideas of laying a groundwork of different documents having different scope selection mechanisms was never fully implemented. Luckily, the existing choice (of whether a document has outputs) happens to be true when we want the single cell scope selection available too, so it would work to just augment the existing area in the code above with a new cell selection scope.

A more ambitious project would be making the find interface more generic to have scope selection for different documents, etc.

**Comment by jess-x** [comment]
@jasongrout thanks so much for the feedback and pointer to right place where this should extend! Having learned about the search scope, the design and the intention in the long run, I completely agree this should be moved to search scope area. I will start working towards that direction. As for making the find interface more generic to specify scope selection per document type, I can try in a separate PR as I think I am still pretty new to the codebase to feel confident at achieving that.

File: packages/documentsearch/src/interfaces.ts (hunk 1 above)
Comment (jess-x) [reply packages/documentsearch/src/interfaces.ts:35]: Thanks for the suggestion, will make this change along with others

**Comment by jasongrout** [comment]
> As for making the find interface more generic to specify scope selection per document type, I can try in a separate PR as I think I am still pretty new to the codebase to feel confident at achieving that.

We appreciate your efforts on this, and definitely, a more ambitious change making things more general can be in a different PR (by you, or by someone else). We don't need to solve all of the problems right now.

Thanks again for working on this.

File: packages/documentsearch/src/searchoverlay.tsx [hunk 2]
```diff
@@ -309,19 +312,38 @@ class FilterSelection extends React.Component<
...
-          checked={this.props.searchOutput}
-          onChange={this.props.toggleOutput}
-        />
+        <div>
+          <span
+            className={
+              this.props.canToggleOutput ? '' : SEARCH_OPTIONS_DISABLED_CLASS
+            }
+          >
+            Search Cell Outputs
```
Comment (goanpeca) [inline packages/documentsearch/src/searchoverlay.tsx:321]: We should probably add localization to this @jess-x

File: packages/documentsearch/src/searchoverlay.tsx [hunk 3]
```diff
@@ -309,19 +312,38 @@ class FilterSelection extends React.Component<
...
+        </div>
+        <div>
+          <span
+            className={
+              this.props.canToggleActiveCell
+                ? ''
+                : SEARCH_OPTIONS_DISABLED_CLASS
+            }
+          >
+            Search Active Cell
```
Comment (goanpeca) [inline packages/documentsearch/src/searchoverlay.tsx:338]: And this!

**Review by goanpeca — CHANGES_REQUESTED** [review]
Awesome work @jess-x !

Couple of requests:
- Could we align checkboxes vertically?
- We should add translation to the strings :)

**Comment by krassowski** [comment]
Great work! 

1) What happens if multiple cells are selected? For me it appears that it still only searches in one "active" cell. I wonder if instead of "Search Active Cell" it should be "Search Selected Cells". As far as I understand, the active cell is always selected, but additional cells can be in the selection too. I would like to see such a generalization, but would not want to block merging this PR as-is otherwise.
- searching/replacing within selected cells is good for feature parity with old notebook
- it was mentioned in three comments in #2017

    Peek 2021-04-18 13-39

2) (minor) while changing active cell, the selection in the previously active cell remains. This is not blocking, but ideally would be addressed at some point because it is visually confusing (and more confusing in certain themes than in others).

    Peek 2021-04-18 13-48

**Review by goanpeca — APPROVED** [review]
Awesome work @jess-x 🚀 !!

Works like a charm

**Comment by jess-x** [comment]
Thanks @krassowski for the feedback!  For the first point you mentioned, I completely agree that's a good parity from classic notebook. I have made the change from active cell to selected cells. For the second point, it's indeed visually confusing but I have not been able to find an easy fix yet. Since it feels non-blocking, I would come back to this later if that's ok.

**Comment by jess-x** [comment]
Hi @jasongrout, I have opened up an issue https://github.com/jupyterlab/jupyterlab/issues/10145 about implementing the general search scope to keep track of it later.

File: packages/documentsearch/src/providers/notebooksearchprovider.ts [hunk 4]
```diff
@@ -359,6 +371,13 @@ export class NotebookSearchProvider implements ISearchProvider<NotebookPanel> {
     return this._changed;
   }
 
+  /**
+   * Signal indicating that the active cell in the notebook has changed
+   */
+  get activeCellChanged(): ISignal<this, void> {
```
Comment (krassowski) [inline packages/documentsearch/src/providers/notebooksearchprovider.ts:377]: I may be missing something, but it appears to me that this signal is not in use (neither emitted or listened to). Is this correct?

Comment (jess-x) [reply packages/documentsearch/src/providers/notebooksearchprovider.ts:377]: Not in use indeed, and I don't think it's needed anymore. Will clean up :) thanks @krassowski

Comment (krassowski) [reply packages/documentsearch/src/providers/notebooksearchprovider.ts:377]: Resolved by #10169. Thanks!

**Pull request merged**